soroban-env-host/src/host.rs

```diff
@@ -325,28 +325,28 @@ impl Host {
         // contract runs out of gas in; this is an atomic action from the
         // contract's perspective.
         let event: DebugEvent = src.into();
-        let len = self.get_events_mut(|events| Ok(events.record_debug_event(event)))?;
-        self.charge_budget(CostType::HostEventDebug, len)
+        self.get_events_mut(|events| {
+            Ok(events.record(InternalEvent::Debug(event), self.as_budget()))
+        })?
     }
 
     // Records a contract event.
-    pub fn record_contract_event(
+    pub(crate) fn record_contract_event(
         &self,
         type_: ContractEventType,
-        topics: ScVec,
-        data: ScVal,
+        topics: Object,
+        data: RawVal,
     ) -> Result<(), HostError> {
-        let ce = ContractEvent {
-            ext: ExtensionPoint::V0,
-            contract_id: self.get_current_contract_id_internal().ok(),
+        self.validate_contract_event_topics(topics)?;
+        let ce = InternalContractEvent {
             type_,
-            body: ContractEventBody::V0(ContractEventV0 { topics, data }),
+            contract_id: self.obj_from_internal_contract_id()?,
```

Comment:
This is a bit of a problem -- it's going to add a new host object on every event. The object for a given contract ID should be added once and cached/reused between events.

Reply:
Shouldn't block this PR though, I've filed https://github.com/stellar/rs-soroban-env/issues/681 to follow up.